### PR TITLE
[bitnami/{common,postgresql}] Error when upgrade with passwords

### PR DESCRIPTION
**Pull request by dani8art**
**Description of the change**

Add helpers to `bitnami/common` chart in order to reuse them for throwing errors when we upgrade chart without required values, like secrets or passwords. 

**Benefits**

W can do upgrade in a safer way. 

**Possible drawbacks**

N/A

**Applicable issues**

N/A

**Additional information**

Also, we have include changes in one of the charts (bitnami/postgresql) so we can test it.

**Checklist** 
- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [x] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

### CODE REVIEW

File: bitnami/common/templates/_errors.tpl [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+{{ include "common.errors.password.empty" (dict "required" (list "mariadb.password" "app.password") "passwords" "instructions" "context" $) }}
+*/}}
+{{- define "common.errors.password.empty" -}}
+{{- if .context.Release.IsUpgrade -}}
+  {{- $validationErrors := include "common.validations.values.empty" . -}}
+
+  {{- if $validationErrors -}}
+    {{- $passwordSection := "" -}}
+    {{- if .passwords -}}
+      {{- $passwordSection = printf "Getting password values:\n\n%s" .passwords -}}
```
Comment (juan131) [inline bitnami/common/templates/_errors.tpl:14]: Do we want to print all these values?

File: bitnami/postgresql/requirements.yaml [hunk 2]
```diff
@@ -1,4 +1,4 @@
 dependencies:
   - name: common
     version: 0.x.x
-    repository: https://charts.bitnami.com/bitnami
\ No newline at end of file
+    repository: https://charts.bitnami.com/bitnami
```
Comment (juan131) [inline bitnami/postgresql/requirements.yaml]: The requirements.lock should be updated, shouldn't it?

File: bitnami/postgresql/templates/secrets.yaml [hunk 3]
```diff
@@ -1,8 +1,14 @@
 {{- if (include "postgresql.createSecret" .) }}
+{{- $requiredPasswords := list "postgresqlPassword" -}}
+{{- $secretName :=  include "postgresql.fullname" . -}}
+
+{{- $getpostgresqlPassword := include "common.notetxt.secret.getvalue" (dict "varname" "POSTGRESQL_PASSWORD" "secret" $secretName "field" "postgresql-password" "context" . ) -}}
+{{- $getPasswords := list $getpostgresqlPassword -}}
+
```
Comment (juan131) [inline bitnami/postgresql/templates/secrets.yaml:7]: I won't include the extra blank lines. Up to you

File: bitnami/common/templates/_notestxt.tpl [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{{/* vim: set filetype=mustache: */}}
+{{/*
+Print instructions to get a secret value.
+Usage:
+{{ include "common.notetxt.secret.getvalue" (dict "varname" "MY_ENVVAR" "secret" "secret-name" "field" "secret-value-field" "context" $) }}
+*/}}
+{{- define "common.notetxt.secret.getvalue" -}}
+export {{ .varname }}=$(kubectl get secret --namespace {{ .context.Release.Namespace }} {{ .secret }} -o jsonpath="{.data.{{ .field }}}" | base64 --decode)
+{{- end -}}
```
Comment (juan131) [inline bitnami/common/templates/_notestxt.tpl:9]: Shouldn't you use this in PostgreSQL's NOTES.txt?

File: bitnami/postgresql/templates/secrets.yaml [hunk 5]
```diff
@@ -11,13 +17,23 @@ metadata:
 type: Opaque
 data:
   {{- if and .Values.postgresqlPostgresPassword (not (eq .Values.postgresqlUsername "postgres")) }}
+    {{- $requiredPassword = append $requiredPassword "postgresqlPostgresPassword" -}}
```
Comment (juan131) [inline bitnami/postgresql/templates/secrets.yaml:20]: `requiredPassword` -> `requiredPasswords`

File: bitnami/common/templates/_notestxt.tpl (hunk 4 above)
Comment (dani8art) [reply bitnami/common/templates/_notestxt.tpl]: done!

File: bitnami/postgresql/templates/secrets.yaml (hunk 5 above)
Comment (dani8art) [reply bitnami/postgresql/templates/secrets.yaml]: N/A

File: bitnami/postgresql/templates/secrets.yaml (hunk 3 above)
Comment (dani8art) [reply bitnami/postgresql/templates/secrets.yaml]: N/A

File: bitnami/postgresql/requirements.yaml (hunk 2 above)
Comment (dani8art) [reply bitnami/postgresql/requirements.yaml]: done

File: bitnami/common/templates/_errors.tpl (hunk 1 above)
Comment (dani8art) [reply bitnami/common/templates/_errors.tpl]: done!

**Comment by dani8art** [comment]
Hi @juan131 I applied your suggestions, please check the new approach. Now when error helm will show

```bash
$ helm upgrade pg bitnami/postgresql
Error: UPGRADE FAILED: template: postgresql/templates/NOTES.txt:72:4: executing "postgresql/templates/NOTES.txt" at <include "common.errors.upgrade.passwords.empty" (dict "required" $requiredPasswords "context" $)>: error calling include: template: postgresql/charts/common/templates/_errors.tpl:21:48: executing "common.errors.upgrade.passwords.empty" at <fail>: error calling fail: 
PASSWORDS ERROR: you must provide your current passwords when upgrade the release
    'postgresqlPassword' must not be empty, please add '--set postgresqlPassword=$POSTGRESQL_PASSWORD' to the command. To get the current value:

        export POSTGRESQL_PASSWORD=$(kubectl get secret --namespace default pg-postgresql -o jsonpath="{.data.postgresql-password}" | base64 --decode)
```

```bash
$ helm upgrade pg bitnami/postgresql --set replication.enabled=true,replication.password=""
Error: UPGRADE FAILED: template: postgresql/templates/NOTES.txt:72:4: executing "postgresql/templates/NOTES.txt" at <include "common.errors.upgrade.passwords.empty" (dict "required" $requiredPasswords "context" $)>: error calling include: template: postgresql/charts/common/templates/_errors.tpl:21:48: executing "common.errors.upgrade.passwords.empty" at <fail>: error calling fail: 
PASSWORDS ERROR: you must provide your current passwords when upgrade the release
    'postgresqlPassword' must not be empty, please add '--set postgresqlPassword=$POSTGRESQL_PASSWORD' to the command. To get the current value:

        export POSTGRESQL_PASSWORD=$(kubectl get secret --namespace default pg-postgresql -o jsonpath="{.data.postgresql-password}" | base64 --decode)

    'replication.password' must not be empty, please add '--set replication.password=$POSTGRESQL_REPLICATION_PASSWORD' to the command. To get the current value:

        export POSTGRESQL_REPLICATION_PASSWORD=$(kubectl get secret --namespace default pg-postgresql -o jsonpath="{.data.postgresql-replication-password}" | base64 --decode)
```

**Comment by dani8art** [comment]
Hi @juan131, I can confirm that lint and test are failing because bitnami/postgresql is unable to get the version of bitnami/common with the required helpers. So I guess we should do in two deferents PR one for the common and another fo postgresql and the rest of charts. 

Please, could you focus on the common helpers and take the postgresql changes as an example of use? and I will separate this PR in two.

Thanks!

**Pull request merged**